racket/collects/racket/dict/contract.rkt

```diff
@@ -0,0 +1,738 @@
+         (define mand-exprs (map in-expr mand-roles mand-tmps))
+         (define opt-exprs (map in-expr opt-roles opt-tmps))
+         (cond
+           [escape-index
+            (define lam-cases
+              (for/list ([i (in-range (add1 (length opt-roles)))])
+                #`[(#,@mand-tmps #,@(take opt-tmps i))
+                   #,(if (and escape-index (<= escape-index i))
+                         #`(let/ec escape
+                             #,(out-expr #`(op #,@mand-exprs #,@(take opt-exprs i))))
+                         (out-expr #`(op #,@mand-exprs #,@(take opt-exprs i))))]))
+            #`(define ((id prj) op)
+                (and op (case-lambda #,@lam-cases)))]
+           [else
+            (define lam-cases
+              (for/list ([i (in-range (add1 (length opt-roles)))])
+                #`[(#,@mand-tmps #,@(take opt-tmps i))
+                   ; escape-index is #false
+                   (values
+                    (λ (ans)
+                      #,(out-expr #'ans))
+                    #,@mand-exprs
+                    #,@(take opt-exprs i))]))
+            #`(define ((id prj) op)
+                (and op
+                     ((if (proj-chaperone-mode? prj)
+                          chaperone-procedure
+                          impersonate-procedure)
+                      op
+                      (case-lambda #,@lam-cases))))]))])))
+
+;; redirect-ref is defined manually,
+;; not with the others in define-dict-redirect-ops,
+;; because the failure-result argument is complicated
+(define ((redirect-ref prj) op)
+  (and
+   op
+   ((if (proj-chaperone-mode? prj)
+        chaperone-procedure
+        impersonate-procedure)
+    op
+    (case-lambda
+      [(self key)
+       (values (λ (ans) (value-out prj ans))
+               self
+               (key-in prj key))]
+      [(self key default)
+       (let ([key (key-in prj key)]
+             [default (project-in prj failure-result/c default)])
+         (cond
+           [(procedure? default)
+            ;; MUTABLE!
+            (define default-called? #f)
+            (define default-result #f)
+            (values (λ (ans)
+                      (cond
+                        [(and default-called? (eq? ans default-result)) ans]
+                        [else (value-out prj ans)]))
+                    self
+                    key
+                    (chaperone-procedure
+                     default
+                     (λ ()
+                       (set! default-called? #t)
+                       (λ (ans)
+                         (set! default-result ans)
+                         ans))))]
+           [else
+            (define default-called?
+              (let/ec default-called!
+                ;; TODO:
+                ;; Is calling the function being chaperoned here evil?
+                ;; It kinda feels evil but I'm not sure.
+                (op self key (λ () (default-called! #t)))
```

Comment:
This line specifically is what feels evil to me. Is it wrong in a procedure chaperone, to call the function I'm supposed to be chaperoning during the chaperone's wrapper-proc? This is what would make it run the user's implementation of `dict-ref` twice for a single call.